design/baremetal-operator/bmh_boot_iso.md

```diff
@@ -0,0 +1,128 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Add boot-iso API to BareMetalHost
+
+## Status
+
+provisional
+
+## Summary
+
+Add a new interface so that it is possible to boot arbitrary
+iso images via Ironic when using redfish-virtualmedia.
+
+## Motivation
+
+In some circumstances it is desirable to boot an existing iso image
+instead of having Ironic boot IPA, for example:
+
+To reduce boot time for ephemeral workloads, it may be possible to boot an iso
+and not deploy any image to disk (saving the time to write the image and reboot)
+
+Where an alternative installer exists that is distributed as a live-iso, it
+may be desirable to leverage that toolchain instead of the IPA deploy ramdisk,
+for example the [fedora-coreos installer](https://docs.fedoraproject.org/en-US/fedora-coreos/bare-metal/)
+
+An [Ironic API](https://docs.openstack.org/ironic/latest/admin/drivers/redfish.html#virtual-media-ramdisk) exists for this, but it's not currently accessible via metal3, so it would be desirable to optionally enable this workflow.
+
+### Goals
+
+Expose the Ironic API to boot iso images via metal3.
+
+### Non-Goals
+
+This only considers the case where the redfish-virtualmedia BMC
+option is used, not any ability to boot arbitrary iso images via
+PXE.
+
+## Proposal
+
+Add a new field to the BareMetalHost spec which indicates that instead of
+deploying an image to disk, a live-iso image will be booted:
+
+```
+  spec:
+    image:
+      iso: http://1.2.3.4/image.iso
+    online: true
+```
+
+Note that in this mode, `rootDeviceHints` and `userData` will not be used
+since the image won't be written to disk, and Ironic doesn't
+support passing user-data in addition to the iso attachment at present.
+
+At some future point it would be desirable for Ironic to add the ability
+to pass userData - then a generic iso could be booted with the ability
+to define customization via the userData field in the normal way.
+
+## Design Details
+
+When this mode is selected, we need to configure Ironic to use the [ramdisk
+deploy interface](https://docs.openstack.org/ironic/stein/admin/interfaces/deploy.html#ramdisk-deploy)
+which is not currently enabled in ironic-image.
```

Comment:
Does that feature use different states in the Ironic state machine? Do we need corresponding changes in the state machine here in metal3, or will all of the changes be within the Ironic provisioner?

Reply:
The state transition is different - as soon as the ISO gets booted, the node is marked ACTIVE, which I think should work without changes to the metal3 state machine?

I did wonder if we should also allow booting ramdisk images via PXE since https://docs.openstack.org/ironic/latest/admin/ramdisk-boot.html supports both modes, and the state machine behavior is the same (then we could support this "live image boot" functionality for all BMC protocols, at the expense of potentially another image type/field)

Reply:
If we support the PXE option, I'm not sure if the `status.image.diskFormat` approach will work well, since we'll need a way to pass both a kernel and ramdisk URL IIUC (that's not explicitly mentioned in the [ironic docs](https://docs.openstack.org/ironic/latest/admin/ramdisk-boot.html#configuration) but I assume we'll need to override the default IPA kernel/ramdisk in driver_info)

Reply:
@dhellmann actually @dtantsur mentioned that Ironic can already boot an ISO image via ipxe, so perhaps using diskFormat would be enough - I've not yet tried that though.

Reply:
Ok to clarify we'd always specify an iso, but internally set a different boot_interface depending on whether it's virtualmedia or not https://docs.openstack.org/ironic/latest/admin/ramdisk-boot.html so if we don't currently see a requirement for separate kernel/ramdisk I can adjust this propoal to specify an iso that can be booted via ipxe or virtualmedia transparently :)

Reply:
I've grown to like the separate `liveBoot` approach, because it feels a bit more explicit about expressing the intent for how to treat the image. I think `diskFormat` will work fine, too, though. I don't think I feel strongly enough about either approach to hold up progress over it.

@zaneb do you have an opinion?

Reply:
I'm fine with `liveBoot` and agree it makes it a little clearer it's a different interface/behavior to the current `image` section.

If we're bikeshedding interfaces I'd perhaps prefer `liveImage` since it aligns slightly better with the existing API but I'm happy to go with the consensus :)

Reply:
I don't know about a checksum but @hardys did mention additional URLs for the ramdisk and kernel. I wouldn't want to add those to `spec.image` because it would be easy to get mixed up about which combination of fields work together and which are ignored in each case. A separate section makes that obvious, even at the expense of possibly having to have "duplicate" fields for some things.

So, how about a `spec.liveImage.url` for now to make it easy to add those other fields later?

Reply:
I like `liveImage.url`.

Reply:
Ok I'll go with `spec.liveImage.url` as that seems to be the consensus, and it leaves the option of adding more fields in future if needed, thanks!